AppCenter/AppCenter/Internals/Context/AuthToken/MSAuthTokenContext.m

```diff
@@ -112,4 +131,89 @@ - (void)removeDelegate:(id<MSAuthTokenContextDelegate>)delegate {
   }
 }
 
+- (nullable NSString *)authToken {
+  NSArray<MSAuthTokenInfo *> *authTokenHistory = [self authTokenHistory];
+  MSAuthTokenInfo *latestAuthTokenInfo = authTokenHistory.lastObject;
+  return latestAuthTokenInfo.authToken;
+}
+
+- (nullable NSString *)accountId {
+  NSArray<MSAuthTokenInfo *> *authTokenHistory = [self authTokenHistory];
+  MSAuthTokenInfo *latestAuthTokenInfo = authTokenHistory.lastObject;
+  return latestAuthTokenInfo.accountId;
+}
+
+- (NSMutableArray<MSAuthTokenValidityInfo *> *)authTokenValidityArray {
+  NSMutableArray<MSAuthTokenInfo *> *__nullable tokenArray =
+      (NSMutableArray<MSAuthTokenInfo *> * __nullable)[MSKeychainUtil arrayForKey:kMSAuthTokenHistoryKey];
+  NSMutableArray<MSAuthTokenValidityInfo *> *resultArray = [NSMutableArray<MSAuthTokenValidityInfo *> new];
+  if (!tokenArray || tokenArray.count == 0) {
+    [resultArray addObject:[[MSAuthTokenValidityInfo alloc] initWithAuthToken:nil startTime:nil endTime:nil]];
+    return resultArray;
+  }
+  for (NSUInteger i = 0; i < tokenArray.count; i++) {
+    MSAuthTokenInfo *currentAuthTokenInfo = tokenArray[i];
+    NSDate *expiresOn = currentAuthTokenInfo.expiresOn;
+    NSDate *nextTokenStartTime = i + 1 < tokenArray.count ? tokenArray[i + 1].startTime : nil;
+    if (nextTokenStartTime && expiresOn && [nextTokenStartTime laterDate:expiresOn]) {
+      expiresOn = nextTokenStartTime;
+    } else if (!expiresOn && nextTokenStartTime) {
+      expiresOn = nextTokenStartTime;
+    }
+    [resultArray addObject:[[MSAuthTokenValidityInfo alloc] initWithAuthToken:currentAuthTokenInfo.authToken
+                                                                    startTime:currentAuthTokenInfo.startTime
+                                                                      endTime:expiresOn]];
+  }
+  return resultArray;
+}
+
+- (void)removeAuthToken:(nullable NSString *)authToken {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *tokenArray = [[self authTokenHistory] mutableCopy];
+
+    // Do nothing if there's just one entry in the history or no history at all.
+    if (!tokenArray || tokenArray.count <= 1) {
+      MSLogWarning([MSAppCenter logTag], @"Couldn't remove token from history; token history is empty or contains only current one.");
+      return;
+    }
+
+    // Check oldest entry, delete if it matches.
+    if ([authToken isEqual:tokenArray[0].authToken]) {
```

Comment:
Channel is flushing log sequentially. Why do you need to check the token is oldest here? Or, why do you take `authToken` value to remove tokens? Seems it is complicated to determine whether a token could be safely removed or not, I would suggest not to delete auth token anywhere except when it needs to add a new one on history full.